src/honey_badger/epoch_state.rs

```diff
@@ -230,54 +263,122 @@ where
-
-                    let faulty_shares: Vec<_> = self
-                        .decryption
-                        .keys()
-                        .filter(|id| !self.accepted_proposers.contains(id))
-                        .cloned()
-                        .collect();
-                    for id in faulty_shares {
-                        if let Some(DecryptionState::Ongoing(td)) = self.decryption.remove(&id) {
-                            for id in td.sender_ids() {
-                                let fault_kind = FaultKind::UnexpectedDecryptionShare;
-                                step.fault_log.append(id.clone(), fault_kind);
-                            }
-                        }
+
+            let Self {
+                accepted_proposers,
+                epoch,
+                decryption,
+                netinfo,
+                subset,
+                subset_handler,
+                ..
+            } = self;
+
+            macro_rules! send_decryption_share {
+                ($proposer_id: expr, $data: expr) => {
+                    step.extend(Self::send_decryption_share(
+                        *epoch,
+                        decryption,
+                        netinfo.clone(),
+                        $proposer_id.clone(),
+                        &$data,
+                    )?);
+                };
+            }
+
+            macro_rules! finalize_subset_with_step {
+                () => {{
+                    has_seen_done = true;
+                    Self::finalize_subset_with_step(subset, accepted_proposers, decryption, &mut step);
+                }};
+            }
+
+            use self::SubsetHandler::*;
+            use self::SubsetOutput::*;
+            match subset_handler {
+                Incremental => match cs_output {
+                    Contribution(proposer_id, data) => {
+                        send_decryption_share!(proposer_id.clone(), &data);
+                        accepted_proposers.insert(proposer_id);
                     }
-                    has_seen_done = true
-                }
+                    Done => finalize_subset_with_step!(),
+                },
+                AllAtEnd(contributions) => match cs_output {
+                    Contribution(proposer_id, data) => {
+                        contributions.push((proposer_id, data));
+                    }
+                    Done => {
+                        accepted_proposers.extend(
+                            contributions
+                                .drain(..)
+                                .map(|(proposer_id, data)| {
+                                    send_decryption_share!(proposer_id.clone(), &data);
+                                    Ok(proposer_id)
+                                }).collect::<Result<Vec<_>>>()?,
+                        );
+
+                        finalize_subset_with_step!();
+                    }
+                },
```

Comment:
Maybe this could be simplified—and the macros, `Self` destructuring and static methods avoided—if `SubsetHandler` had a `handle` method that takes a `SubsetOutput` and returns a `Vec<SubsetOutput>`:
* If it's `Incremental`, it returns its argument immediately.
* If it's `AllAtEnd` it stores each `Contribution` input, and then returns all of them later, when `Done` arrives.
We wouldn't need an explicit distinction here, and the code could remain almost unchanged.

Reply:
That makes sense -- I thought about doing that and decided to accept the complexity to avoid heap allocations and copying. You definitely are more familiar with what tradeoffs the codebase needs, though, so I'll make that change once I get to a dev machine.

Reply:
I think that would be premature optimization: The bottleneck in hbbft is on way lower layers (mainly in the cryptography), so our priority should be to keep the code simple.